### ISSUE TITLE
Implementation of the translation of the enumerations.

### ISSUE THREAD

**Issue by stefsava** [marmelab/admin-on-rest#455]
In order to translate the enumerated fields I've implemented EnumField component below. You feel that it was the right way to approach the problem?

I would do the same for Inputs. I tried unsuccessfully to implement a similar solution as ReferenceInput so you can provide the choices property SelectInput, RadioButtonGroupInput, AutocompleteInput etc.

What do you think?

```
// src/mui/field/EnumField.js

import React, { PropTypes } from 'react';
import pure from 'recompose/pure';
import { translate } from 'admin-on-rest';
import Chip from 'material-ui/Chip';

/**
 * @example
 * <EnumField source="gender" />
 * in i18n messages you need to add
 * { en: {resources: { people: { enums: { gender: { m: 'Male', f: 'Female' } } } } } }
 */

const EnumField = ({ record, source, resource, elStyle = { margin: 4 }, translate }) => (
  <Chip style={elStyle} >{translate(`resources.${resource}.enums.${source}.${record[source]}`)}</Chip>
);

EnumField.propTypes = {
  addLabel: PropTypes.bool,
  elStyle: PropTypes.object,
  label: PropTypes.string,
  resource: PropTypes.string.isRequired,
  record: PropTypes.object.isRequired,
  source: PropTypes.string.isRequired,
  translate: PropTypes.func.isRequired,
};

const PureEnumField = pure(EnumField);

PureEnumField.defaultProps = {
  addLabel: true,
};

export default translate(PureEnumField);

```

**Comment by fzaninotto** [comment]
I'm not sure I understood your question. Let me rephrase what I understood.

1. You ask the we add a new component for displaying translated enum values in the core. Is it a relevant addition? Is the suggested implementation good enough?
2. You're thinking about a way to let a user choose a translated enum value, but you didn't manage to. You're asking for help.

To these questions, I'd reply:

1. Yes, it is a good idea, but I don't agree with the implementation. To me, we should pass the enum text directly to `translate()`, something like:

```js
/**
 * @example
 * <EnumField source="gender" choices={[
 *    { id: 'm', name: 'Male' },
 *    { id: 'f', name: 'Female' },
 * ]}/>
 * in i18n messages you need to add
 * { en: {resources: { people: { enums: { gender: { m: 'Male', f: 'Female' } } } } } }
 */
const EnumField = ({ record, source, choices, elStyle, translate }) => {
  const choice = get(record, source);
  const label = choices.find(c => c.id === choice);
  return <Chip style={elStyle} >{translate(label)}</Chip>
};
const PureEnumField = pure(EnumField);

PureEnumField.defaultProps = {
  addLabel: true,
};

export default translate(PureEnumField);
```

You can observe a certain similarity with the `<SelectInput>` syntax. That's also the answer to your question #2: we don't need to add new components, we just need to translate the labels displayed by `<SelectInput>`, `<AutocompleteInput>` and `< RadioButtonGroupInput>`. But that must be optional, as it doesn't make sense to translate values in `<ReferenceInput>`. So I'd suggest an API like:

```js
/**
 * @example
 * <EnumField source="gender" choices={[
 *    { id: 'm', name: 'Male' },
 *    { id: 'f', name: 'Female' },
 * ]}/>
 * in i18n messages you need to add
 * { en: {resources: { people: { enums: { gender: { m: 'Male', f: 'Female' } } } } } }
 */
const EnumField = ({ record, source, choices, elStyle, translate, translateLabels = true }) => {
  const choice = get(record, source);
  const label = choices.find(c => c.id === choice);
  return <Chip style={elStyle} >{translateLabels ? translate(label) : label}</Chip>
};
const PureEnumField = pure(EnumField);

PureEnumField.defaultProps = {
  addLabel: true,
};

export default translate(PureEnumField);
```

**Comment by stefsava** [comment]
First, to avoid misunderstandings, I'm not referring to the translation of the label, where "label" we mean the name of the field, eg the header of the columns of a list. In the code above constant name "label" might confuse other readers.

The code posted by @fzaninotto is fine for enums that can be shared among multiple resources, such as for example the proposed "gender".

The translation files must be for your example like this:
```
export default {
  it: {
    gender: {
      m: 'Male',
      f: 'Female',
    }
  }
}
```
In the case in which enums are typical of a given resource I believe is appropriate to include them within the branch of the resource.
```
export default {
  en: {
    resources: {
      projects: {
        name: 'Project',
        fields: {
          description: 'Description',
          status: 'Status',
        },
        enums: {
          status: {
            not_set: 'Not Started',
            request_start: 'In Queue',
            working: 'Ongoing',
            paused: 'On Hold',
          },
        },
      },
    },
  },
};
```
In this case I find it boring and less maintainable make it compulsory to specify the choices properties.

```
export default props => (
  <List {...props} >
    <Datagrid>
      <NumberField source="id" />
      <TextField source="description" />
      <EnumField
        source="status"
        choices={[
          { id: 'not_set', name: 'resources.projects.enums.status.not_set' },
          { id: 'request_start', name: 'resources.projects.enums.status.request_start' },
          { id: 'working', name: 'resources.projects.enums.status.working' },
          { id: 'paused', name: 'resources.projects.enums.status.paused' },
        ]}
      />
      <EditButton />
      <DeleteButton />
      <ShowButton />
    </Datagrid>
  </List>
);

```
It seems more DRY have a more smart component capable of generating the right keys able to find the correct translations and able to make choices an optional property
```
export default props => (
  <List {...props} >
    <Datagrid>
      <NumberField source="id" />
      <TextField source="description" />
      <EnumField source="status" />
      <EditButton />
      <DeleteButton />
      <ShowButton />
    </Datagrid>
  </List>
);
```

I agree about not adding too many components in the main project "admin-on-rest".

I could publish a separate module with EnumField and EnumInput.
I'm sure aor is expected to have a large following of addons.

**Comment by fzaninotto** [comment]
Again, I think "more DRY [...] a more smart" also means "more magic & more WTF" for developers. With your implementation, they would need to read the doc to understand where the translations should end up. With mine, they decide. I think mine is better.

I'll probably add the Enum components to the core. If you want to appear as the author, feel free to send a PR.

**Comment by fzaninotto** [comment]
Fixed by #501
